jupyter_server_fileid/manager.py

```diff
@@ -104,6 +104,39 @@ def _from_normalized_path(self, path: Optional[str]) -> Optional[str]:
 
         return relpath
 
+    def _move_recursive(self, old_path: str, new_path: str, sep: str = os.path.sep) -> None:
```

Comment:
Nice.  I like how directories can be handled implicitly like this.

Comment: 
Since persisted paths are always normalized to absolute, OS-specific, paths, is there a need to specify the path separator since it should always be `os.path.sep`?  I think the `ArbitraryFileIdManager`'s methods will have problems on Windows where it unconditionally passes a separator of `"/"`.  I don't think the additional parameter is necessary.

Hmm - I'm beginning to wonder if api paths should be retained and persisted as api paths irrespective of OS - at least within the `ArbitraryFileIdManager`.  For example, if I have a `ContentsManager` that happens to use "/" to walk (i.e., as a separator of "tables") a NoSQL database, and that database is available for use by multiple servers configured with the same `ContentsManager` (which is likely very common for database-related CMs), then servers running Windows will be persisting entirely different paths into the `Files` table than servers running Linux/MacOS and, well, things will break.

Also, `_normalize_path()` has a bug in that it retains the complete api path when joining `os.path.join(self.root_dir, path)`.   This will result in a Windows path of `C:\\my\\root\\dir\\my/api/path`.

FWIW, jupyter_server.utils contains two methods that may be helpful: [`to_api_path()`](https://github.com/jupyter-server/jupyter_server/blob/122758fe8983a839c9883e9d745fec254aa2fbc0/jupyter_server/utils.py#L117) and [`to_os_path()`](https://github.com/jupyter-server/jupyter_server/blob/122758fe8983a839c9883e9d745fec254aa2fbc0/jupyter_server/utils.py#L105).

Reply:
> Since persisted paths are always normalized to absolute, OS-specific, paths, is there a need to specify the path separator since it should always be os.path.sep? I think the ArbitraryFileIdManager's methods will have problems on Windows where it unconditionally passes a separator of "/". I don't think the additional parameter is necessary.

Great catch, since it allows me to elaborate more on the current implementation. The relative API path we receive will always be specified with forward slashes. The only time we are using `os.path.sep` in ArbitraryFileIdManager is to join the root directory with the relative API path we get. So on Windows, given a contents manager root of `C:\\root`, the persisted path in the database will look something like this:

```
C:\\root\\some/path/file
```

So when dealing with directories underneath the root, the children are always delimited by forward slashes. Hence we need to explicitly pass the `sep = '/'` argument. This is the "bug" you were mentioning later in your comment.

But why did I choose to do it this way? I agree that it is odd that on Windows that we are using `\\` as a path separator to join the root dir to the path, but the portion after the root dir is using `/`. The reason I chose to not call `os.path.normpath()` on the API path we receive is because:

1. We don't need to do it, since we're never accessing the local filesystem in ArbitraryFileIdManager anyways.

2. It leads to some weird issues if the ArbitraryFileIdManager is running on Windows, but the ContentsManager is giving us a relative path in which `\` is a perfect valid character for a file name. [Take S3, which allows `\` to be part of a filename (though it discourages its use)](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html). For example, consider the file ID manager receiving an API path for `a\\b/c`. Here the contents manager is telling us there is a directory named `a\\b` with a child `c`. Then upon translation, the path we store in the record is:

```
s3://bucket\\a\\b\\c
```

Here, we can see that the information that `a\\b` is actually a file name is completely lost. We now think `a` is a directory containing `b` as a child. Moving a file named `a` will move this file, even though it shouldn't. Whereas if we leave the portion after the root dir of the path untouched and assume forward slashes to always separate children (current implementation), the path we store in the record is:

```
s3://bucket\\a\\b/c
```

Thus ArbitraryFileIdManager correctly handles the case where a custom ContentsManager implementation gives us file names with `\\` even if it's invalid on the current OS.

The case is different for LocalFileIdManager, which has a dependency on using the FileContentsManager. Here we can assume that the file paths are all local, so given that FileContentsManager is running on the same machine, `\\` will never be used as part of a file name on Windows, so it is always safe to translate the API path via `os.path.normpath()`. That's why we override `_normalize_path()` for LocalFileIdManager.

I get that this can be confusing (I spent all day yesterday thinking about this). I'm happy to go over this again if necessary. 😅 

> For example, if I have a ContentsManager that happens to use "/" to walk (i.e., as a separator of "tables") a NoSQL database, and that database is available for use by multiple servers configured with the same ContentsManager (which is likely very common for database-related CMs), then servers running Windows will be persisting entirely different paths into the Files table than servers running Linux/MacOS and, well, things will break.

They shouldn't break, since users of FileIdManager should never be accessing the path stored in the database. That's an implementation detail that shouldn't be relied on; they should only get paths via `get_path()`, which will always return an API path delimited by forward slashes regardless of the OS the ArbitraryFileIdManager is running on.

Reply:
Ah crap. This still is wrong if you have a local contents manager and change the root on Windows.

```
# root = c:\\root
_normalize_path("a/b") # => "c:\\root\\a/b

# root = c:\\root\\a
_normalize_path("b") # => "c:\\root\\a\\b
```

I'll need to rethink the implementation again to handle this edge case. Sorry about that lengthy write up that was ultimately wrong.